### PR TITLE
gitlab: use custom git

### PR DESCRIPTION
**Pull request by ajs124**
## Description of changes
for improved compatibility

not super polished, but we've been running something similar for a long time, internally

## Things done
- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ajs124** [comment]
turns out, they've dropped `libgit2`, so maybe lets wait with this for a release that doesn't need it anymore?

https://gitlab.com/gitlab-org/gitaly/-/merge_requests/6317

**Comment by ajs124** [comment]
https://github.com/NixOS/nixpkgs/pull/256484 is said release, without libgit2. so once that (or something similar) is merged, I'll rework this and undraft.

**Review by yayayayaka — COMMENTED** [review]
Sorry for leaving this open for so long.

With the mentioned suggestions applied locally, I have deployed the changes for manual testing. git clone, git pull, git push, creating and merging merge requests works, so I think this should be good to go.

File: nixos/modules/services/misc/gitlab.nix [hunk 1]
```diff
@@ -1130,7 +1131,7 @@ in {
       }
     ];
 
-    environment.systemPackages = [ pkgs.git gitlab-rake gitlab-rails cfg.packages.gitlab-shell ];
+    environment.systemPackages = [ git gitlab-rake gitlab-rails cfg.packages.gitlab-shell ];
```
[the remark this answers is not in the document]
Comment (ajs124) [reply nixos/modules/services/misc/gitlab.nix:1134]: But we're already passing `pkgs.git` in `systemPackages` now? So just remove it from there?

File: nixos/modules/services/misc/gitlab.nix
[the remark this answers is not in the document]
Comment (ajs124) [reply nixos/modules/services/misc/gitlab.nix:1]: fixed

**Comment by teto** [comment]
could you elaborate in the commit message "for improved compatibility" with what ? what's the problem with default git ?

File: nixos/modules/services/misc/gitlab.nix [hunk 2]
```diff
@@ -176,16 +178,15 @@ let
     MALLOC_ARENA_MAX = "2";
   } // cfg.extraEnv;
 
-  runtimeDeps = with pkgs; [
```
Comment (dasJ) [inline nixos/modules/services/misc/gitlab.nix:179]: You can do this in the gitlab package:
```
  # This 11 should be more or less in sync with https://gitlab.com/gitlab-org/gitlab/-/blob/master/lib/gitlab.rb#L36
  postPatch = ''
    echo '${lib.substring 0 11 data.rev}' > REVISION
  '';
```
and remove git from all runtime inputs except gitlab-config :smirk:

**Pull request closed without merging**